File: 2019/day/4/Main.hs
import Data.List

increasing len = go len 1
  where
    go 0   _ = [[]]
    go len d = do
        x <- [d..9]
        xs <- go (len - 1) x
        return (x:xs)

main = do
    let input = "146810-612564"
        range@(lo, hi) = ([1,4,6,8,1,0], [6,1,2,5,6,4])

    -- Part 1
    print $ length
          $ filter (any ((>= 2) . length) . group)
          $ takeWhile (<= hi)
          $ dropWhile (<= lo)
          $ increasing 6

    -- Part 2
    print $ length
          $ filter (any ((== 2) . length) . group)
          $ takeWhile (<= hi)
          $ dropWhile (<= lo)
          $ increasing 6

File: 2019/day/4/main.py
#%%
from itertools import groupby
lo, hi = map(int, "146810-612564".split('-', 2))

#%%
# Part 1
%time sum(any([len(list(g)) >= 2 for k, g in groupby(s, lambda x: x)]) \
          for s in map(str, range(lo, hi + 1)) \
          if all(map(lambda x: x[0] <= x[1], zip(s, s[1:]))))
 
# Part 2
#%%
%time sum(any([len(list(g)) == 2 for k, g in groupby(s, lambda x: x)]) \
          for s in map(str, range(lo, hi + 1)) \
          if all(map(lambda x: x[0] <= x[1], zip(s, s[1:]))))